lib/hl7/parser.ex

```diff
@@ -0,0 +1,118 @@
+defmodule HL7.Parser do
+  @carriage_return 4
+  @field 3
+  @repetition 2
+  @component 1
+  @sub_component 0
+
+  def parse(<<"MSH|^~\\&", rest::binary()>> = text) do
+    tokenize(rest, text, 8, 0, ["^~\\&", @field, "|", @field, "MSH"])
+    |> to_lists()
+  end
+
+  defp tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
```

Comment:
I'm finding the indirect recursion between `tokenize` and `tokenize_terminator` a little hard to follow. I was thinking you could probably do the same work with a fold over the characters and no explicit recursion. I had to try it before I suggested it, and it seems to work for me. I know the focus of this PR is performance, which might be affected by any tweak to the approach, so I'm going to try to get `Benchee` results on the fold.

Reply:
Argh, the fold made it slower! My world is shattered. Who cares about performance anyway?

Reply:
I took a shot at inlining `tokenize_terminator` to make the recursion easier to follow. See what you think about this.
```elixir
  defp tokenize(<<char::binary-size(1), rest::binary>>, original, skip, len, acc) do
    case get_terminator(char) do
      nil ->
        tokenize(rest, original, skip, len + 1, acc)

      terminator ->
        new_acc =
          if len == 0 do
            acc
          else
            [binary_part(original, skip, len) | acc]
          end

        tokenize(rest, original, skip + len + 1, 0, [terminator | new_acc])
    end
  end

  defp get_terminator(character) do
    case character do
      "|" -> @field
      "~" -> @repetition
      "^" -> @component
      "&" -> @sub_component
      "\r" -> @carriage_return
      "\n" -> @carriage_return
      _ -> nil
    end
  end
```

Reply:
that looks really nice!!! :) does it retain performance that way?

Reply:
Yes, it does! I need to make Benchee a regular habit. Performance metrics almost always defy my intuitions.

Reply:
alas, i've found that unrolling things even farther gave me _big_ speed boosts so I have to table this :(